File: answers/lintcode/Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param head a ListNode
     * @param k an integer
     * @return a ListNode
     */
    ListNode *reverseKGroup(ListNode *head, int k) {
        // Write your code here
        ListNode *dummy = new ListNode(INT_MAX);
        int count = 0;
        while (count < k && head) {
            ListNode *temp = head->next;
            head->next = dummy->next;
            dummy->next = head;
            head = temp;
            ++ count;
        }
        if (count < k && INT_MAX != k) {
            dummy->next = reverseKGroup(dummy->next, INT_MAX);
        }
        if (head) {
            ListNode *temp = dummy;
            while (temp && temp->next) {
                temp = temp->next;
            }
            if (temp) {
                temp->next = reverseKGroup(head, k);
            }
        }
        return dummy->next;
    }
};

File: answers/leetcode/Reverse Nodes in k-Group/Reverse Nodes in k-Group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode current = head;
        for (int i = 1; i < k && current != null; i++) {
            current = current.next;
        }
        if (current != null) {
            current = head;
            for (int i = 1; i < k; i++) {
                dummy.next = head.next;
                head.next = head.next.next;
                dummy.next.next = current;
                current = dummy.next;
            }
            head.next = this.reverseKGroup(head.next, k);
        }
        return dummy.next;
    }
}